Mathlib/RingTheory/Polynomial/Bernstein.lean

```diff
@@ -0,0 +1,419 @@
+#align bernstein_polynomial.map bernsteinPolynomial.map
+
+end
+
+theorem flip (n ν : ℕ) (h : ν ≤ n) :
+    (bernsteinPolynomial R n ν).comp (1 - X) = bernsteinPolynomial R n (n - ν) := by
+  simp [bernsteinPolynomial, h, tsub_tsub_assoc, mul_right_comm]
+#align bernstein_polynomial.flip bernsteinPolynomial.flip
+
+theorem flip' (n ν : ℕ) (h : ν ≤ n) :
+    bernsteinPolynomial R n ν = (bernsteinPolynomial R n (n - ν)).comp (1 - X) := by
+  simp [← flip _ _ _ h, Polynomial.comp_assoc]
+#align bernstein_polynomial.flip' bernsteinPolynomial.flip'
+
+theorem eval_at_0 (n ν : ℕ) : (bernsteinPolynomial R n ν).eval 0 = if ν = 0 then 1 else 0 := by
+  rw [bernsteinPolynomial]
+  split_ifs with h
+  · subst h; simp
+  · simp [zero_pow (Nat.pos_of_ne_zero h)]
+#align bernstein_polynomial.eval_at_0 bernsteinPolynomial.eval_at_0
+
+theorem eval_at_1 (n ν : ℕ) : (bernsteinPolynomial R n ν).eval 1 = if ν = n then 1 else 0 := by
+  rw [bernsteinPolynomial]
+  split_ifs with h
+  · subst h; simp
+  · obtain w | w := (n - ν).eq_zero_or_pos
+    · simp [Nat.choose_eq_zero_of_lt ((tsub_eq_zero_iff_le.mp w).lt_of_ne (Ne.symm h))]
+    · simp [zero_pow w]
+#align bernstein_polynomial.eval_at_1 bernsteinPolynomial.eval_at_1
+
+theorem derivative_succ_aux (n ν : ℕ) :
+    Polynomial.derivative (bernsteinPolynomial R (n + 1) (ν + 1)) =
+      (n + 1) * (bernsteinPolynomial R n ν - bernsteinPolynomial R n (ν + 1)) := by
+  rw [bernsteinPolynomial]
+  suffices ((n + 1).choose (ν + 1) : R[X]) * ((↑(ν + 1 : ℕ) : R[X]) * X ^ ν) * (1 - X) ^ (n - ν) -
+      ((n + 1).choose (ν + 1) : R[X]) * X ^ (ν + 1) * ((↑(n - ν) : R[X]) * (1 - X) ^ (n - ν - 1)) =
+      (↑(n + 1) : R[X]) * ((n.choose ν : R[X]) * X ^ ν * (1 - X) ^ (n - ν) -
+        (n.choose (ν + 1) : R[X]) * X ^ (ν + 1) * (1 - X) ^ (n - (ν + 1))) by
+    simpa [Polynomial.derivative_pow, ← sub_eq_add_neg, Nat.succ_sub_succ_eq_sub,
+      Polynomial.derivative_mul, Polynomial.derivative_nat_cast, MulZeroClass.zero_mul,
+      Nat.cast_add, algebraMap.coe_one, Polynomial.derivative_X, mul_one, zero_add,
+      Polynomial.derivative_sub, Polynomial.derivative_one, zero_sub, mul_neg, Nat.sub_zero,
+      bernsteinPolynomial, map_add, map_natCast, Nat.cast_one]
+  conv_rhs => rw [mul_sub]
+  -- We'll prove the two terms match up separately.
+  refine' congr (congr_arg Sub.sub _) _
+  · simp only [← mul_assoc]
+    refine' congr (congr_arg (· * ·) (congr (congr_arg (· * ·) _) rfl)) rfl
+    -- Now it's just about binomial coefficients
+    exact_mod_cast congr_arg (fun m : ℕ => (m : R[X])) (Nat.succ_mul_choose_eq n ν).symm
+  · rw [← tsub_add_eq_tsub_tsub, ← mul_assoc, ← mul_assoc]; congr 1
+    rw [mul_comm, ← mul_assoc, ← mul_assoc]; congr 1
+    norm_cast
+    congr 1
+    convert (Nat.choose_mul_succ_eq n (ν + 1)).symm using 1
+    · rw [mul_comm, Nat.succ_sub_succ_eq_sub]
+      -- Porting note: was
+      -- convert mul_comm _ _ using 2
+      -- simp
```

Comment:
```suggestion
    · -- Porting note: was
      -- convert mul_comm _ _ using 2
      -- simp
      rw [mul_comm, Nat.succ_sub_succ_eq_sub]
```
Not essential, but I find it less confusing when the porting notes are consistently *before* the material they discuss.